File: source/com/barrybecker4/puzzle/twopails/ui/TwoPailsRenderer.java
/** Copyright by Barry G. Becker, 2013. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.puzzle.twopails.ui;

import com.barrybecker4.puzzle.common.PuzzleRenderer;
import com.barrybecker4.puzzle.twopails.model.PailParams;
import com.barrybecker4.puzzle.twopails.model.Pails;

import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;

/**
 * Singleton class that takes a PieceList and renders it for the TwoPailsViewer.
 * Having the renderer separate from the viewer helps to separate out the rendering logic
 * from other features of the TwoPailsViewer.
 *
 * @author Barry Becker
 */
public class TwoPailsRenderer implements PuzzleRenderer<Pails> {

    public static final float CONTAINER_WIDTH = 0.2f;
    public static final float SEPARATION = 0.11f;
    public static final int TEXT_WIDTH = 70;
    public static final int TEXT_OFFSET = 10;

    private static final int TOP_MARGIN = 160;
    private static final int MARGIN = 60;

    private static final Color CONTAINER_COLOR = new Color(5, 0, 80);
    private static final Color LIQUID_COLOR = new Color(95, 145, 255);

    private static final Font FONT = new Font("Sans Serif", Font.BOLD, 16);


    /**
     * private constructor because this class is a singleton.
     * Use getPieceRenderer instead.
     */
    public TwoPailsRenderer() {}

    /**
     * This renders the current state of the Pails to the screen.
     */
    @Override
    public void render( Graphics g, Pails pails, int width, int height) {

        PailParams params = pails.getParams();
        byte biggest = params.getBiggest();
        float size1 = (float)params.getPail1Size() / biggest;
        float size2 = (float)params.getPail2Size() / biggest;

        int usableWidth = width - MARGIN;
        int contWidth = (int)(CONTAINER_WIDTH * usableWidth);
        int cont1Height = (int)(size1 * (height - TOP_MARGIN));
        int cont2Height = (int)(size2 * (height - TOP_MARGIN));

        int middle = MARGIN + TEXT_WIDTH + (int)((CONTAINER_WIDTH + SEPARATION) * usableWidth);
        int container1Y = height - MARGIN - cont1Height;
        int container2Y = height - MARGIN - cont2Height;

        g.setColor(Color.BLACK);
        g.setFont(FONT);
        g.drawString("First Pail", MARGIN, MARGIN + TEXT_OFFSET);
        g.drawString("Max  = " + params.getPail1Size(), MARGIN, container1Y + TEXT_OFFSET );
        g.drawString("Fill = " + pails.getFill1(), MARGIN, height - MARGIN);

        g.drawString("Second Pail", middle, MARGIN + TEXT_OFFSET);
        g.drawString("Max = " + params.getPail2Size(), middle, container2Y + TEXT_OFFSET);
        g.drawString("Fill = " + pails.getFill2(), middle, height - MARGIN);

        // show outlines for two containers
        g.setColor(CONTAINER_COLOR);
        g.drawRect(MARGIN + TEXT_WIDTH, container1Y, contWidth, cont1Height);
        g.drawRect(middle + TEXT_WIDTH, container2Y, contWidth, cont2Height);

        g.setColor(LIQUID_COLOR);
        // show fill for first container
        int fillHeight = (int)((float)pails.getFill1()/biggest * cont1Height) - 2;
        g.fillRect(MARGIN + TEXT_WIDTH + 1, height - MARGIN - fillHeight, contWidth-1, fillHeight);

        // show fill for second container
        fillHeight = (int)((float)pails.getFill2()/biggest * cont1Height) - 2;
        g.fillRect(middle + TEXT_WIDTH + 1, height - MARGIN - fillHeight, contWidth-1, fillHeight);
    }

}

File: scala-source/com/barrybecker4/puzzle/twopails/ui/TwoPailsRenderer.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.twopails.ui

import com.barrybecker4.puzzle.common.PuzzleRenderer
import com.barrybecker4.puzzle.twopails.model.Pails
import java.awt.Color
import java.awt.Font
import java.awt.Graphics


/**
  * Singleton class that takes a PieceList and renders it for the TwoPailsViewer.
  * Having the renderer separate from the viewer helps to separate out the rendering logic
  * from other features of the TwoPailsViewer.
  *
  * @author Barry Becker
  */
object TwoPailsRenderer {
  private val CONTAINER_WIDTH = 0.2f
  private val SEPARATION = 0.11f
  private val TEXT_WIDTH = 70
  private val TEXT_OFFSET = 10
  private val TOP_MARGIN = 160
  private val MARGIN = 60
  private val CONTAINER_COLOR = new Color(5, 0, 80)
  private val LIQUID_COLOR = new Color(95, 145, 255)
  private val FONT = new Font("Sans Serif", Font.BOLD, 16)
}

class TwoPailsRenderer extends PuzzleRenderer[Pails] {

  /** This renders the current state of the Pails to the screen. */
  override def render(g: Graphics, pails: Pails, width: Int, height: Int): Unit = {
    val params = pails.getParams
    val biggest = params.getBiggest
    val size1 = params.pail1Size.toFloat / biggest
    val size2 = params.pail2Size.toFloat / biggest
    val usableWidth = width - TwoPailsRenderer.MARGIN
    val contWidth = (TwoPailsRenderer.CONTAINER_WIDTH * usableWidth).toInt
    val cont1Height = (size1 * (height - TwoPailsRenderer.TOP_MARGIN)).toInt
    val cont2Height = (size2 * (height - TwoPailsRenderer.TOP_MARGIN)).toInt
    val middle = TwoPailsRenderer.MARGIN + TwoPailsRenderer.TEXT_WIDTH +
      ((TwoPailsRenderer.CONTAINER_WIDTH + TwoPailsRenderer.SEPARATION) * usableWidth).toInt
    val container1Y = height - TwoPailsRenderer.MARGIN - cont1Height
    val container2Y = height - TwoPailsRenderer.MARGIN - cont2Height

    g.setColor(Color.BLACK)
    g.setFont(TwoPailsRenderer.FONT)
    g.drawString("First Pail", TwoPailsRenderer.MARGIN, TwoPailsRenderer.MARGIN + TwoPailsRenderer.TEXT_OFFSET)
    g.drawString("Max  = " + params.pail1Size, TwoPailsRenderer.MARGIN, container1Y + TwoPailsRenderer.TEXT_OFFSET)
    g.drawString("Fill = " + pails.fill1, TwoPailsRenderer.MARGIN, height - TwoPailsRenderer.MARGIN)
    g.drawString("Second Pail", middle, TwoPailsRenderer.MARGIN + TwoPailsRenderer.TEXT_OFFSET)
    g.drawString("Max = " + params.pail2Size, middle, container2Y + TwoPailsRenderer.TEXT_OFFSET)
    g.drawString("Fill = " + pails.fill2, middle, height - TwoPailsRenderer.MARGIN)

    // show outlines for two containers
    g.setColor(TwoPailsRenderer.CONTAINER_COLOR)
    g.drawRect(TwoPailsRenderer.MARGIN + TwoPailsRenderer.TEXT_WIDTH, container1Y, contWidth, cont1Height)
    g.drawRect(middle + TwoPailsRenderer.TEXT_WIDTH, container2Y, contWidth, cont2Height)
    g.setColor(TwoPailsRenderer.LIQUID_COLOR)
    // show fill for first container
    var fillHeight = (pails.fill1.toFloat / biggest * cont1Height).toInt - 2
    g.fillRect(TwoPailsRenderer.MARGIN + TwoPailsRenderer.TEXT_WIDTH + 1, height - TwoPailsRenderer.MARGIN - fillHeight, contWidth - 1, fillHeight)
    // show fill for second container
    fillHeight = (pails.fill2.toFloat / biggest * cont1Height).toInt - 2
    g.fillRect(middle + TwoPailsRenderer.TEXT_WIDTH + 1, height - TwoPailsRenderer.MARGIN - fillHeight, contWidth - 1, fillHeight)
  }
}


